### ISSUE TITLE
The Realm is already in a write transaction	

### ISSUE THREAD

**Issue by feidaZhang** [realm/realm-js#6834]
### How frequently does the bug occur?

Sometimes

### Description

# context
We found the issue in realm 10.15.0 and `react-native` 0.66.3 one year ago.
After we upgrade `realm` to 12.6.2 and `react-native` to 0.72.12. We still can get these exceptions from logs in production.
**However, we can not repro it on our side.** It's too bad.
We call the following method in a loop. I am not sure if it causes the issue.
```  
public static saveTransaction(entity) {
    try {
      RealmManager.getRealmInstance().write(() => {
        RealmManager.getRealmInstance().create('Transaction', entity, UpdateMode.All);
      });
      return true;
    } catch (exception) {
      return false;
    }
  }
```
## errors on iOS 17.5.1 
The exceptions happen frequently to the user, who is on iOS 17.5.1. 
![realm_error_ios](https://github.com/user-attachments/assets/1cbbbfff-1498-4396-aa81-33ce63eb41d4)

![image](https://github.com/user-attachments/assets/cfdc08b0-0617-42f8-80e5-b816ac61f795)

## errors on other users
issue: The Realm is already in a write transaction
**Also, there are a few error logs and they are all instantaneous in other Android and iOS users.**
![image](https://github.com/user-attachments/assets/8d7953e5-0183-49aa-9d59-5b46e2cbbed5)

### Can you reproduce the bug?

No

### Version

12.6.2

### What services are you using?

Local Database only

### Are you using encryption?

No

### Platform OS and version(s)

iOS 17.5.1

### Build environment

Which debugger for React Native: ..

**Comment by kneth** [comment]
It is difficult to tell as the functionality of `RealmManager.getRealmInstance()` is unknown. My guess is that the method returns a cached Realm instance which is already in a write transaction.

Instead of `write()`, you can use manual transactions: [`Realm.isInTransaction()`](https://www.mongodb.com/docs/realm-sdks/js/latest/classes/Realm-1.html#isInTransaction), [`Realm.beginTransaction()`](https://www.mongodb.com/docs/realm-sdks/js/latest/classes/Realm-1.html#beginTransaction), [`Realm.commitTransaction()`](https://www.mongodb.com/docs/realm-sdks/js/latest/classes/Realm-1.html#commitTransaction), and [`Realm.cancelTransaction()`](https://www.mongodb.com/docs/realm-sdks/js/latest/classes/Realm-1.html#cancelTransaction).

**Comment by feidaZhang** [comment]
Yes. We have a cached realm singleton globally. However, this instance will not be modified after it is created.
All our business operations run on the JS thread of RN, why does this still cause this problem? @kneth 
and, may i know the difference between `write()` and `beginTransaction(), commitTransaction()` ? because we always use the `write()`.
```
let realm: Realm;
export default class RealmManager {
  static getRealmInstance = () => {
    return realm;
  };
}```

**Comment by kneth** [comment]
You could try to write your method to: 

```js

public static saveTransaction(entity) {
    const realm = RealmManager.getRealmInstance();
    let inTransaction = realm.isInTransaction;
    if (!inTransaction) {
      realm.beginTransaction();
    }
    realm.create('Transaction', entity, UpdateMode.All);
    if (!inTransaction) {
      realm.commitTransaction();
    }
  }
```

**Comment by feidaZhang** [comment]
> You could try to write your method to:
> 
> ```js
> public static saveTransaction(entity) {
>     const realm = RealmManager.getRealmInstance();
>     let inTransaction = realm.isInTransaction;
>     if (!inTransaction) {
>       realm.beginTransaction();
>     }
>     realm.create('Transaction', entity, UpdateMode.All);
>     if (!inTransaction) {
>       realm.commitTransaction();
>     }
>   }
> ```

Do you mean we are able to reuse the current transaction?

**Comment by kneth** [comment]
`mmap() failed: Cannot allocate memory ...` - it sounds like you have a transaction which is too large to handle for the database. Can you split it multiple transactions?
